charon/src/bin/charon-driver/translate/translate_types.rs

```diff
@@ -779,7 +781,101 @@ impl<'tcx, 'ctx> ItemTransCtx<'tcx, 'ctx> {
     }
 }
 
+fn translate_variant(
+    variant_layout: &r_abi::LayoutData<r_abi::FieldIdx, r_abi::VariantIdx>,
+) -> VariantLayout {
+    match &variant_layout.fields {
+        r_abi::FieldsShape::Primitive
+        | r_abi::FieldsShape::Union(_)
+        | r_abi::FieldsShape::Array { .. } => VariantLayout::default(),
+        r_abi::FieldsShape::Arbitrary { offsets, .. } => {
+            let mut v = Vector::with_capacity(offsets.len());
+            for (i, o) in offsets.iter_enumerated() {
+                // The fields should have a total increasing index, but to be
+                // absolutely sure, we insert them at the correct index.
+                v.insert(FieldId::from_usize(i.index()), o.bytes());
+            }
+            VariantLayout { field_offsets: v }
+        }
+    }
+}
+
 impl ItemTransCtx<'_, '_> {
+    /// Translate a type layout.
+    ///
+    /// Translates the layout as queried from rustc into
+    /// the more restricted [`Layout`].
+    #[tracing::instrument(skip(self))]
+    pub fn translate_layout(&self) -> Layout {
+        let tcx = self.t_ctx.tcx;
+        let rdefid = self.def_id.as_rust_def_id().unwrap();
+        let ty_env = hax::State::new_from_state_and_id(&self.t_ctx.hax_state, rdefid).typing_env();
+        let ty = tcx.type_of(rdefid).skip_binder();
```

Comment:
I'm careful about `skip_binder`s, because they tend to be mistakes. This one is ok so I'm adding a comment.
```suggestion
        // This `skip_binder` is ok because it's an `EarlyBinder`.
        let ty = tcx.type_of(rdefid).skip_binder();
```